### PR TITLE
Add function for running retriable promises

### PR DESCRIPTION
**Pull request by fjvallarino**
@tejohnso @ahmedalsudani please review

### CODE REVIEW

File: platform.js [hunk 1]
```diff
@@ -423,5 +423,16 @@ module.exports = {
         });
       }
     });
+  },
+  runPromise(promise, retries) {
+    return promise()
+      .catch((err)=>{
+        if(retries > 0) {
+          return module.exports.runPromise(promise, retries - 1);
```
Comment (ahmedalsudani) [inline platform.js]: @fjvallarino can we add a timeout here? As for the timeout interval, I'm not sure what number we should go with, but we should make it take long enough so the display has a chance to reconnect (not necessarily on the first attempt, but after a few attempts)

Comment (fjvallarino) [inline platform.js]: @ahmedalsudani makes sense. I'll add it as an optional parameter. Should we retry on timeouts or just give up?

**Comment by fjvallarino** [comment]
@tejohnso @ahmedalsudani updated

File: platform.js [hunk 2]
```diff
@@ -423,5 +423,52 @@ module.exports = {
...
+        var timer;
+
+        if(retryTimeout) {
+          timer = setTimeout(()=>{
+            handleError("function call timed out");
+          }, retryTimeout);
+        }
+
+        func()
+          .then(clearTimer)
```
Comment (tejohnso) [inline platform.js]: So func has to be a function but it also has to be a function that returns a promise. We should guard for that, or maybe coerce to a Promise via `Promise.resolve(func()).then...`

**Comment by tejohnso** [comment]
Wow that really grew.

**Comment by fjvallarino** [comment]
@tejohnso @ahmedalsudani updated

**Comment by tejohnso** [comment]
:+1:

File: platform.js [hunk 3]
```diff
@@ -423,5 +423,67 @@ module.exports = {
...
+          .then(resolve.bind(null, errors))
+          .catch(handleError);
+
+        function handleError(err) {
+          clearTimer();
+
+          errors.push(err);
+
+          if(retries === 0) {
```
Comment (ahmedalsudani) [inline platform.js]: @fjvallarino I think this should be `<= 0` (even though that can only happen on invalid input)

**Comment by ahmedalsudani** [comment]
Once my comment above is addressed, :+1:

File: platform.js (hunk 3 above)
Comment (fjvallarino) [inline platform.js]: @ahmedalsudani I don't really think we need that

Comment (ahmedalsudani) [inline platform.js]: @fjvallarino then we should validate that `retries` is valid when we enter the function so we don't keep retrying forever

**Pull request merged**